### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by SSE4**
- split implementations into their own files
- added Circle CI
- added AZP (Azure DevOps Pipelines)
- by default, read config files from `XDG_CONFIG_HOME`
- add verbose logging for API requests

### CODE REVIEW

**Comment by uilianries** [comment]
Seems like we have a big change here! Thanks @SSE4 awesome contribution, I'll review it on Sunday.

**Review by Croydon — COMMENTED** [review]
I have installed this version, created an AZP token with full access rights and executed `bincrafters-envy --skip-travis --skip-appveyor --with-azure -c "myfilepath\bincraftersenvy.ini" -p "conan-openscenegraph"`. I was expecting that openscenegraph has now an added pipeline here https://dev.azure.com/bincrafters/packages/_build but this isn't the case 🤔 

[Edit] Output:

```
bincrafters-envy --skip-travis --skip-appveyor --with-azure -c "myfilepath\bincraftersenvy.ini" -p "conan-openscenegraph"
INFO:root:args: myfilepath\bincraftersenvy.ini
INFO:root:using config: myfilepath\bincraftersenvy.ini
INFO:root:updating project conan-openscenegraph on azure...
WARNING:root:project conan-openscenegraph already exists on azure
INFO:root:updating project conan-openscenegraph on azure...OK
```

[Edit2] Okay, I guess it isn't supposed to add new pipelines yet, right?

However, we actually don't need to add credentials for every single pipeline, as one variable group shared for all pipelines is enough.

So it would be nice if we can figure a way out to add new pipelines programmatically.

**Comment by SSE4** [comment]
@Croydon but it's here: https://dev.azure.com/bincrafters/packages/_build
image

**Comment by Croydon** [comment]
Sorry, you are right.

Pipelines without runs aren't displayed in the list view and I wasn't aware that this folder view exists 

azp

**Comment by Croydon** [comment]
Still doesn't seem to be correctly set as the repository / ressource information is missing

https://dev.azure.com/bincrafters/packages/_apps/hub/ms.vss-ciworkflow.build-ci-hub?_a=edit-build-definition&id=44&view=Tab_Tasks

azprepo

**Comment by SSE4** [comment]
@Croydon interesting, I'll take a look

**Comment by SSE4** [comment]
@Croydon I've got a little bit further, now all properties are filled
the missing part is how to set up service connection with GitHub via API (which isn't created automatically)
```
Could not find the service connection associated with this pipeline. You can use the classic editor to fix the service connection in the YAML source settings.
```
if I click to edit pipeline manually, connection is automatically added by Web GUI...

**Comment by SSE4** [comment]
update: seems to be working with hard-coded end-point
there is an API to create one: https://docs.microsoft.com/en-us/rest/api/azure/devops/serviceendpoint/endpoints/create?view=azure-devops-rest-5.1
some code for reference: https://github.com/Azure/azure-devops-cli-extension/blob/ae2d49831f11601e44b30257e9d23334638fd770/azure-devops/azext_devops/dev/pipelines/pipeline_create_helpers/pipelines_resource_provider.py#L17
I'll try to implement it a bit later

File: bincrafters_envy/appveyor.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    def add_one(self, project_slug):
+        repository_name = '{accountName}/{projectSlug}'.format(
+            accountName=self._github_account,
+            projectSlug=project_slug
+        )
+        request = {
+            'repositoryProvider': 'gitHub',
+            'repositoryName': repository_name
+        }
+        self._post(url='/projects', data=json.dumps(request))
```
Comment (uilianries) [inline bincrafters_envy/appveyor.py]:
```suggestion
        self._post(url='/projects', json=request)
```
So you can remove `Application/Content-Type` from headers

File: bincrafters_envy/appveyor.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            name = v['name']
+            if name not in new_env_vars.keys():
+                new_env_vars[name] = v
+
+        request = []
+
+        for _, v in new_env_vars.items():
+            request.append(v)
+
+        self._put(url=appveyor_url, data=json.dumps(request), expected_status=204)
```
Comment (uilianries) [inline bincrafters_envy/appveyor.py]:
```suggestion
        self._put(url=appveyor_url, json=request, expected_status=204)
```

File: bincrafters_envy/appveyor.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    def _encrypt(self, value):
+        appveyor_url = '/account/encrypt'
+        request = {'plainValue': value}
+        encrypted = self._post(url=appveyor_url, data=json.dumps(request))
```
Comment (uilianries) [inline bincrafters_envy/appveyor.py]:
```suggestion
        encrypted = self._post(url=appveyor_url, json=request)
```

File: bincrafters_envy/azure.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        self._endpoint = self._host + '/' + self._account + "/packages/_apis"
+        self._api_version = "?api-version=5.0"
+        self._headers = {'Content-type': 'application/json'}
+        self._auth = HTTPBasicAuth(username=self._token, password="")
+
+    def _get_default_branch(self, url):
+        import subprocess
+        import tempfile
+
+        # is there a better way, e.g. without clone?
```
Comment (uilianries) [inline bincrafters_envy/azure.py:27]: If the project is available on Github, you use their REST API to collect this information. Also, Conan tools provides a smooth helper to Git commands. Anyway, your solution works fine.

File: bincrafters_envy/azure.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        data["repository"]["id"] = "%s/%s" % (self._account, project_slug)
+        data["repository"]["name"] = "%s/%s" % (self._account, project_slug)
+        data["repository"]["type"] = "GitHub"
+        data["repository"]["url"] = github_url
+        data["repository"]["defaultBranch"] = self._get_default_branch(github_url)
+        data["repository"]["clean"] = "false"
+        data["repository"]["checkoutSubmodules"] = False
+        data["repository"]["properties"] = {'connectedServiceId': '1b6eeb61-8694-4b38-8804-96e4340e4660'}
+
+        self._post(url=url, data=json.dumps(data))
```
Comment (uilianries) [inline bincrafters_envy/azure.py]:
```suggestion
        self._post(url=url, json=data)
```

File: bincrafters_envy/azure.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        for name, value in env_vars.items():
+            data["variables"][name] = {
+                "value": value,
+                "isSecret": name in encrypted_vars
+            }
+        groups = self._get(url=url)
+        if groups["count"] > 0:
+            groupid = groups["value"][0]["id"]
+            url = "/distributedtask/variablegroups/" + str(groupid) + self._api_version + "-preview.1"
+            self._put(url=url, data=json.dumps(data))
```
Comment (uilianries) [inline bincrafters_envy/azure.py]:
```suggestion
            self._put(url=url, json=data)
```

File: bincrafters_envy/azure.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        else:
+            url = "/distributedtask/variablegroups" + self._api_version + "-preview.1"
+            self._post(url=url, data=json.dumps(data))
```
Comment (uilianries) [inline bincrafters_envy/azure.py]:
```suggestion
            self._post(url=url, json=data)
```

File: bincrafters_envy/base.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    def _delete(self, **kwargs):
+        return self._request(method="DELETE", **kwargs)
+
+    def _patch(self, **kwargs):
+        return self._request(method="PATCH", **kwargs)
+
+    def _put(self, **kwargs):
+        return self._request(method="PUT", **kwargs)
+
+    def _request(self, method, url, data=None, expected_status=200):
```
Comment (uilianries) [inline bincrafters_envy/base.py]:
```suggestion
    def _request(self, method, url, json=None, expected_status=200):
```

File: bincrafters_envy/base.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        url = self._endpoint + url
+        r = requests.request(method=method, url=url, headers=self._headers, auth=self._auth, data=data)
```
Comment (uilianries) [inline bincrafters_envy/base.py]:
```suggestion
        r = requests.request(method=method, url=url, headers=self._headers, auth=self._auth, json=data)
```

File: bincrafters_envy/base.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        if r.status_code != expected_status:
+            raise Exception('%s %s request failed %s %s' % (self.name, method, r.status_code, r.content))
+        content_type = r.headers.get("Content-Type") or ""
+        if "application/json" in content_type:
+            # charmap' codec can't encode character '\u015f' in position 169832: character maps to <undefined>
```
Comment (uilianries) [inline bincrafters_envy/base.py]: nick trick, I didn't know about.

File: bincrafters_envy/base.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+            return json.loads(r.content.decode('utf-8', 'replace'))
```
Comment (uilianries) [inline bincrafters_envy/base.py:65]:
```suggestion
            return r.json()
```
this does not work?

File: bincrafters_envy/bincrafters_envy.py [hunk 12]
```diff
@@ -1,308 +1,115 @@
...
-    print('[y/n]')
-    choice = input().lower()
-    while choice not in ['y', 'n']:
-        print('please respond with y or n')
-        choice = raw_input().lower()
-    return choice == 'y'
-
+__author__ = "BinCrafters"
+__license__ = "MIT"
+__version__ = '0.1.7'
```
Comment (uilianries) [inline bincrafters_envy/bincrafters_envy.py]:
```suggestion
__version__ = '0.2.0'
```
Big change requires a new minor ;)

File: bincrafters_envy/circle.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        url = self._project_url(project_slug) + "/envvar"
+        old_vars = self._get(url=url)
+        for var in old_vars:
+            name = var["name"]
+            if name not in env_vars:
+                self._delete(url=url + "/" + name)
+
+        for k, v in env_vars.items():
+            request = {"name": k, "value": v}
+            self._post(url=url, data=json.dumps(request), expected_status=201)
```
Comment (uilianries) [inline bincrafters_envy/circle.py]:
```suggestion
            self._post(url=url, json=request, expected_status=201)
```

File: bincrafters_envy/travis.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+        for name, value in env_vars.items():
+            request = {
+                'env_var.name': name,
+                'env_var.value': value,
+                'env_var.public': name not in encrypted_vars
+            }
+            if name in ids.keys():
+                travis_url_env = self._project_url(project_slug) + '/env_var/' + ids[name]
+                self._patch(url=travis_url_env, data=json.dumps(request))
```
Comment (uilianries) [inline bincrafters_envy/travis.py]:
```suggestion
                self._patch(url=travis_url_env, json=request)
```

File: bincrafters_envy/travis.py [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+            else:
+                self._post(url=travis_url, data=json.dumps(request), expected_status=201)
```
Comment (uilianries) [inline bincrafters_envy/travis.py]:
```suggestion
                self._post(url=travis_url, json=request, expected_status=201)
```

**Comment by uilianries** [comment]
I need to test locally, but this change is really cool, big refactoring and new support. Thanks @SSE4

**Comment by uilianries** [comment]
Travis CI .org = Okay with my user

Appveyor = 500 Internal error. I'm curious if is related to my account.

First, I tried to add my conan-glib fork to Appveyor, but an error occurred:

[envy-conan-glib.log](https://github.com/bincrafters/bincrafters-envy/files/3563512/envy-conan-glib.log)


So I tried to update a Conan Catch, my fork from Bincrafters, some error:

[envy-conan-catch.log](https://github.com/bincrafters/bincrafters-envy/files/3563515/envy-conan-catch.log)

File: bincrafters_envy/azure.py (hunk 4 above)
Comment (SSE4) [reply bincrafters_envy/azure.py:27]: yeah, I anyway will need to obtain GitHub token from the AZP in order to establish service connection

File: bincrafters_envy/base.py (hunk 11 above)
Comment (SSE4) [reply bincrafters_envy/base.py:65]: no idea, need to check, because some json from appveyor contained incorrect utf-8 sequences

**Pull request closed without merging**